google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleFrameworkDetectorTest.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.facet.flexible;
+
+import com.intellij.mock.MockVirtualFile;
+import com.intellij.openapi.fileTypes.FileTypeRegistry;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.patterns.ElementPattern;
+import com.intellij.util.indexing.FileContent;
+import com.intellij.util.indexing.FileContentImpl;
+import org.jetbrains.yaml.YAMLFileType;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/** Tests {@link AppEngineFlexibleFrameworkDetector} for valid/invalid App Engine Flex detection. */
+public class AppEngineFlexibleFrameworkDetectorTest {
+
+  private final String validAppEngineFlexYamlString =
+      AppEngineFlexibleFrameworkDetector.APP_ENGINE_REQUIRED_YAML + " java";
```

Comment:
Instead of using the constant in the file, use the actual value you would expect to pass. For example, change this to be "runtime: java".

This separation between test and code under test is important for several reasons. One is that if this constant were changed to something unexpected (i.e. a bug), the test would pass in its current form. However, if it's hardcoded to the expected value of "runtime: java" in the test, the developer would have to specifically change it in the test too, at which point they might realize the constant should not have changed.

It's a bit of a stretch in this case since how often do constants "accidentally" change? But consider the case where the field used contains logic (i.e. a `Function`, `Consumer`, etc.) or the shared piece of code is actually a method? It's easier to see how the test no longer validates the code under test does what it's supposed to do since it shares code with it.